### DISCUSSION TITLE
Extending ndarray for binary arrays

### DISCUSSION THREAD

**Discussion by AdamWRichardson** [rust-ndarray/ndarray#1459] [Q&A]
Hi everyone,

I want to create a new crate which extends on the great ndarray work but is specific for binary matrices. My two new types I'd like to create are:
* Bitpacked binary array - where the underlying data is a bitvec or bitarray (from [this crate](https://docs.rs/bitvec/1.0.1/bitvec/index.html))
* Sparse binary array - which could have many different implementations

I was hoping to get some pointers on the best way to achieve this. Thanks!

**Comment by akern40** [comment]
Hello! On the topic of bitpacked arrays, this has been brought up in #1258 but wasn't implemented. Is this the same idea - essentially a `BitArrD`? On the topic of sparse representations, we currently don't really have a way to switch out the underlying memory representation under `ndarray`, nor a set of traits that could be implemented for anything that is _like_ an `ndarray`. Both of these things are of interest to me, however.

**Reply by AdamWRichardson** [reply]
Thanks for the reply, good spot with that issue although a shame about the proprietary implementation. Is there already a proposal for a trait based memory API, or a discussion about how to implement it? I think that would be incredibly useful to have not just for my use cases but also to enable testing and bench-marking of different implementations (for example when [std simd](https://doc.rust-lang.org/std/simd/index.html) is stabilised).

**Reply by akern40** [reply]
There isn't an official RFC - you can check out [the design-doc branch of akern40/ndarray-design](https://github.com/akern40/ndarray-design/tree/design-doc), where I have a document that lays out one potential design for a new base type that allows for customized backends (among other things). I'm frankly unsatisfied with that particular design, but there are some difficult pros and cons to weigh between a trait-based vs type-based design for switchable backends.
